source/common/http/conn_manager_impl.cc

```diff
@@ -987,6 +967,16 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(RequestHeaderMapPtr&& he
   ASSERT(!cached_route_);
   refreshCachedRoute();
 
+  // Note: this is the earliest point at which we can record headers size stats, given that
+  // cluster info becomes available after refreshCachedRoute() is called. So any additional
+  // headers that were added before this point will be accounted for, thus this number will
+  // be bigger than expected.
+  Upstream::ClusterRequestResponseSizeStatsOptRef req_resp_stats =
+      clusterRequestResponseSizeStats();
+  if (req_resp_stats.has_value()) {
+    req_resp_stats->get().upstream_rq_headers_size_.recordValue(request_headers_->byteSize());
+  }
+
```

Comment:
This will record the stats for the cluster selected before any of the filters ran, which doesn't seem ideal: the target cluster might change before it hits the router, which would misattribute the values

Any way to do this after we know the upstream for sure? Can this be done in the router? Or log this once the stream ends, reading the upstream cluster from the stream info?

Reply:
Hmm, would doing it from completeRequest() -- where we record values for req/resp body size -- work?

Reply:
Alright we chatted offline, we'll push all the recording into completeRequest() and note the limitations of this approach (doesn't work well with direct responses, retries, etc).

Reply:
Had 2nd thoughts, let's get it right -- I'll move this to the router.